lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,450 @@
+defmodule Scholar.Cluster.Hierarchical do
+  @moduledoc """
+  Performs [hierarchical, agglomerative clustering](https://en.wikipedia.org/wiki/Hierarchical_clustering#Agglomerative_clustering_example)
+  on a dataset.
+
+  Hierarchical clustering is good for when the number of clusters is not known ahead of time.
+  It also allows for the creation of a [dendrogram plot](https://en.wikipedia.org/wiki/Dendrogram)
+  (regardless of the dimensionality of the dataset) which can be used to select the number of
+  clusters in a post-processing step.
+
+  <!-- (This isn't ready yet.)
+
+  ## Examples
+
+  See the [_Hierarchical Clustering_](./doc/notebooks/hierarchical_clustering.livemd) Livebook for
+  a worked example with a dendrogram plot.
+
+  -->
+
+  ## Limitations
+
+  Due to the requirements of the current implementation, only these options are supported:
+
+    * `dissimilarity: :euclidean`
+    * `linkage: :average | :complete | :single | :ward | :weighted`
+
+  Our current algorithm is $O(\\frac{n^2}{p})$ where $n$ is the number of data points
+  and $p$ is the number of processors.
```

Comment:
Looking back over what I've got, I'm not totally sure what the complexity is. I think each clade-finding pass is $O(\frac{n^2}{p})$, but I'm still getting used to Nx.

Reply:
[9a2db56](https://github.com/elixir-nx/scholar/pull/187/commits/9a2db56753949f7e3faff55331bed4b28e9d426f)